pallets/tinlake-investor-pool/src/lib.rs

```diff
@@ -890,28 +890,37 @@ pub mod pallet {
 		fn validate_pool_constraints(
 			reserve: T::Balance,
 			max_reserve: T::Balance,
-			tranche_ratios: &[Perquintill],
+			min_risk_buffers: &[Perquintill],
 			current_tranche_values: &[T::Balance],
 		) -> DispatchResult {
-			if tranche_ratios.len() != current_tranche_values.len() {
+			if min_risk_buffers.len() != current_tranche_values.len() {
 				Err(Error::<T>::InvalidData)?
 			}
 
 			if reserve > max_reserve {
 				Err(Error::<T>::InsufficientReserve)?
 			}
 
-			for (count, tv) in current_tranche_values.iter().enumerate() {
-				let acc_sub_value = current_tranche_values
-					.iter()
-					.skip(count + 1)
-					.fold(Zero::zero(), |sum: T::Balance, tvs| {
-						sum.checked_add(tvs).unwrap()
-					});
-
-				let calc_sub: Perquintill = Perquintill::from_rational(acc_sub_value, *tv);
-				if calc_sub < tranche_ratios[count] {
-					Err(Error::<T>::SubordinationRatioViolated)?
+			let total_value = current_tranche_values
+				.iter()
+				.fold(
+					Some(Zero::zero()),
+					|sum: Option<T::Balance>, tranche_value| {
+						sum.and_then(|sum| sum.checked_add(tranche_value))
+					},
+				)
+				.ok_or(Error::<T>::Overflow)?;
+			let mut buffer_value = total_value;
+			for (tranche_value, min_risk_buffer) in current_tranche_values
+				.iter()
+				.zip(min_risk_buffers.iter().copied())
+			{
+				buffer_value = buffer_value
+					.checked_sub(tranche_value)
+					.ok_or(Error::<T>::Overflow)?;
+				let risk_buffer = Perquintill::from_rational(buffer_value, total_value);
+				if risk_buffer < min_risk_buffer {
+					Err(Error::<T>::RiskBufferViolated)?
```

Comment:
How do we ensure we are not getting in this state?

Reply:
This is used as part of validating solutions (and should be renamed or at least documented as such). The solution submitted will need to properly limit the investments and redemptions of the different tranches to keep the constraints satisfied.